### ISSUE TITLE
exclusive field in the jobspec

### ISSUE THREAD

**Issue by dongahn** [flux-framework/rfc#101]
Hi @morrone: Two more things on the jobspec. 

1) It looks like your current jobspec implementation doesn't yet handle "exclusive" field. I noticed a "Fix me" tag in the source code so I believe it is in your todo list.

2) Also, I think we should be a bit careful when this support is added. The RFC currently specifies

> The exclusive key SHALL be a boolean indicating, when true, that the current resource is requested to be allocated exclusively to the current program. If unset, the default value for exclusive SHALL be false for vertices that are not within a task slot. The default value for exclusive SHALL be true for task slots (type: slot) and their associated resources.

Here, in addition to the explicit slot type, I believe the implicit slot-serving resource type (i.e., slot only showing up in the Task section pointing to the real resource type in the resource section) should also have "exclusive" set to be true by default. It would be nice if this is clarified in the RFC.

**Comment by morrone** [comment]
Sure, I can add exclusive soon.

I'm not sure that exclusive/nonexlusive has any meaning for slots.  Perhaps it would be better to just add wording that makes it clear that that field has no meaning for slots?  I could make it an error to use it with slots too.

**Comment by dongahn** [comment]
From resource matching point of view, having "exclusive=true" set for both explicit and implicit slot types will be significantly helpful. Essentially, when I see "exclusive=true" at a resource vertex, I can propagate the exclusivity to all of the  resources at the subtree.

**Comment by morrone** [comment]
I guess that the current spec does say "The default value for exclusive SHALL be true for task slots (type: slot) and their associated resources."  But it seems a bit odd to make "exclusive" apply recursively below slot, but not apply recursively below other resource types.

If we leave it the way it is, I think that implies that I need more than a bool value for "exclusive", because "unspecified" is a valid state, and it is interpreted differently depending on where it appears in the tree.

**Comment by dongahn** [comment]
> I guess that the current spec does say "The default value for exclusive SHALL be true for task slots (type: slot) and their associated resources."

Yes, it does say it. But it would be clearer if we say this includes "implicit slot types." To make my point a bit more concrete, let's look at one of the current jobspec examples:

```yaml
version: 1
resources:
    - type: cluster
      count: 2
      with:
          - type: node
            count: 15
            with:
              - type: core
                count: 1
tasks:
  - command: [ "flux", "start" ]
    slot:
      type: node
    count:
      per_slot: 1
attributes:
  system:
    duration: 1 hour
```

We don't have an explicit slot type in the resource section. Bu my resource matcher should interpret `node` type as a slot type (i.e., implicit since the user did not explicitly specify the `slot` type). So, if "exclusive" field is already set to be "true" by your jobspec implementation for both implicit as well as explicit slot types, this will be helpful.


> If we leave it the way it is, I think that implies that I need more than a bool value for "exclusive", because "unspecified" is a valid state, and it is interpreted differently depending on where it appears in the tree.

I agree. Perhaps, we should specify all of the resource types under the slot types (explicit/implicit) shall also be exclusive?

This can be a bit funky if we later add edge types, but we agreed that we won't worry about this for now.

**Comment by morrone** [comment]
> We don't have an explicit slot type in the resource section. Bu my resource matcher should interpret node type as a slot type (i.e., implicit since the user did not explicitly specify the slot type)

That is not clear to me from reading the specification.  It seems to me that it only applies when there is an explicit resource vertex of type "slot".  But it is not really even clear to me that explicit slots type should change sharing behaviour...I would think that we want the scheduler to mostly decide sharing based on default settings rather than what the jobspec says, unless the user explicitly requests exclusive in the jobspec.

For instance, in the example that you gave, I see no reason why that jobspec should be interpreted as the user asking for nodes to be assigned explicitly.  I would think that node default exclusivity should be decided in the configuration of the scheduler.  For instance, nodes are assigned exclusively on most of our clusters by default, but on other clusters, nodes are shared.  If the jobspec interprets a slot at a "node" to automatically make the node request exclusive, doesn't that mean that nodes will almost never be shared on a shared node cluster?  We would have to get all of the users to explicitly say "explicit=false" to go back to sharing nodes.

I think this is why I left exclusive as "FIX ME".  I wasn't fully understanding the spec as written, and wasn't sure what would be handled where.

As a first pass, I think I'll update libjobspec to use a tri-state value for the exclusive field: FALSE, TRUE, and UNSPECIFIED.  That keeps libjobspec's job as being a simpler parser/translater, rather that it attempting to determine any actually meaning from the contents of a jobspec.  Doing any more would require me to move to a more complicated multi-pass parser.  I would like to avoid that if possible.  (Then again, right now I am using yaml-cpp which forces me to work in two passes anyway, so maybe it wouldn't be too hard to do as you ask.  But first we should decide that it really is a good way to go.)

**Comment by morrone** [comment]
I've been mulling over SLURM's "overcommit" option and how something similar (and perhaps better) might be implemented in Flux.

I am thinking if the jobspec is mostly interpretted as "UNSPECIFIED" at parsing time, that leaves us the flexibility at scheduling time to do some interesting things with overcommit or resource without having to custom edit each jobspec submission.

For instance, a user could tell their personal flux instance's scheduler that it wants to allow up to 8 tasks to be explicitly launched per core.  Then if there is a very large number of jobspecs that require a single task on a single core, the scheduler can know that it can run 8 jobspecs per core.

It allows a level of overcommit, without making overcommit runaway without limit.

And again, if task slot resources (whether implicity or explicit) default to "exclusive=true", overcommit would probably not be possible.

**Comment by grondo** [comment]
> For instance, in the example that you gave, I see no reason why that jobspec should be interpreted as the user asking for nodes to be assigned explicitly. I would think that node default exclusivity should be decided in the configuration of the scheduler. For instance, nodes are assigned exclusively on most of our clusters by default, but on other clusters, nodes are shared. If the jobspec interprets a slot at a "node" to automatically make the node request exclusive, doesn't that mean that nodes will almost never be shared on a shared node cluster? 

I've only been half following this discussion, so let me apologize in advance if my statements confuse or take us off topic.

On first read, I would have also assumed that when a slot is specified by a jobspec author, it implies that all the resources within the slot are exclusively required. Thus, on a shared node cluster, nodes would be shared because users will be requesting slots that use a subset of what is on a node, such a single core or a single socket.

I was initially confused by @dongahn's example above, where a slot is defined in the task section. It does make sense however (I think) if you define the "node" slot relative to the resource request, which appears to be asking for 15 cores, each on a different node. In this case the "exclusive" slot is not the whole node, but all the resources assigned on each node.

For whatever it is worth, your idea to use "unspecified" for exclusive if the state is not specified seems good to me.

**Comment by grondo** [comment]
> I've been mulling over SLURM's "overcommit" option and how something similar (and perhaps better) might be implemented in Flux.

Yeah, that overcommit thing always seemed a bit antiquated. Why can't a user, once they have been assigned resources, run as many processes as they want on those resources?

Any reason it needs to be handled specially? Can't we allow this in the `tasks` section by allowing the user to run more than one defined task per slot? e.g. `per_slot: 20`?

> For instance, a user could tell their personal flux instance's scheduler that it wants to allow up to 8 tasks to be explicitly launched per core. Then if there is a very large number of jobspecs that require a single task on a single core, the scheduler can know that it can run 8 jobspecs per core.

I do like this idea (again, for whatever it is worth). However, once the user launches a task on a core, it gets difficult to prevent that task from launching more processes, so I'm not sure how hard you'd want to try to "share" cores in this way....

**Comment by morrone** [comment]
>  It does make sense however (I think) if you define the "node" slot relative to the resource request, which appears to be asking for 15 cores, each on a different node.

Actually, since the the tasks section says "slot: type: node", I believe that the the node itself also becomes exclusive according to the current spec (plus @dongahn 's additional request that implicit slots obey the same rules as exclicit slots).

It does result in a strange situation in which the job may be allocated a node with 32 cores, but only 1 core on each node is actually accessible to the job.  So perhaps that is another reason that we should think about changing what the rfc says.

**Comment by grondo** [comment]
> It does result in a strange situation in which the job may be allocated a node with 32 cores, but only 1 core on each node is actually accessible to the job. So perhaps that is another reason that we should think about changing what the rfc says.

Agreed. While the RFC says what it says, I'm pretty confident the *intent* was for slot definitions in the task section to be relative to the resource request (i.e., once the request is satisfied, in this case you have 15 nodes with 1 core and associated memory, etc. THEN you define the slot relative to that, so the slot is a node and 1 core (and associated memory))

In fact, I think early on I think I remember us saying the scheduler might not even be required to pay attention to slots (unless they are required to get the full shape of the job request), since that is more an execution system job (layout of tasks onto resources). Sounds like that has changed so a new pass through the RFC is warranted.

**Comment by dongahn** [comment]
Ok I chatted with @morrone a bit. Here are some of the action items.

1. We should be more specific about what "implicit slot" means in our RFC. I mistakenly thought an implicit slot  has the same effect as having the slot type right above the referred resource type, but it is actually right below the referred type. So the above example has the same effect as:

```yaml
version: 1
resources:
    - type: cluster
      count: 2
      with:
          - type: node
            count: 15
            with:
            - slot: slot
               count: 1
               label: here
               with:
                  - type: core
                    count: 1
tasks:
  - command: [ "flux", "start" ]
    slot:
      label: here
    count:
      per_slot: 1
attributes:
  system:
    duration: 1 hour

```

This is okay as far as it is clearly defined in the RFC. Maybe it is; it may just be me not seeing these definitions.


2.  The scheduler needs to pay attention to slots

From talking to @morrone and @lipari the other day, we found some cases where the scheduler needs to pay attention to `slot` for valid resource grouping. So, I still need to add this support within my current effort. 

3. Exclusivity with `slot` type

It seems we need to think about this a bit more. As @morrone and I were talking this through, however, it was clear this can become quickly complex. Also, we were not sure what was the justification and use cases to make `slot` and resource below exclusive in the first place anyway.

It seems like there are many benefits if each scheduler can have its own policy on exclusivity and etc on certain resource types. But I can't wait until we flesh out all the details. (There is a much bigger fish to fry for me.) 

So, while we will continue to discuss this later down the road, what I will do is to only support a limited job specs (specs with explicit exclusivity associated and slot's exclusivity has the current semantics) for testing purposes to make progress with my current efforts.

**Comment by morrone** [comment]
> Any reason it needs to be handled specially? Can't we allow this in the tasks section by allowing the user to run more than one defined task per slot? e.g. per_slot: 20?

You are correct that within a single job that would work.  I wasn't at all clear, but I was thinking more about the ability to oversubscribe resources across multiple jobs in a controlled way.  So for instance you have 10,000 1-task jobspec files and your application gets the best throughput by running four of those tasks per core, then flux might be able to schedule those jobs nice and evenly in a way that slurm can't very easily do.  The scheduler would just need to be told that 4-way overcommit of the cores is permitted.

Or maybe a new resource type named "quartercore" is created instead for the subinstance, and then the jobspecs all ask for 1 quartercore.  I don't know.  Lots of ways to skin a cat with flux. :)

**Comment by dongahn** [comment]
On a second thought about the implicit slot:

Given that one of the design goals of the canonical job spec is:

> Emphasize expressivity over simplicity, as this canonical form may be generated from other user-friendly forms or interfaces.

Why do we even introduce the implicit slot? Unless there is a compelling reason for this implicitly, can we disallow implicit slots? This will guarantee the resource section in @morrone's jobspec implementation will always have "slot type", which will help reduce the complexity of schedulers.

**Comment by dongahn** [comment]
If this is not all that clear, my hope is to make it so that the schedulers don't even have to look at the sections other than resources section.

**Comment by grondo** [comment]
> Why do we even introduce the implicit slot? Unless there is a compelling reason for this implicitly, can we disallow implicit slots? This will guarantee the resource section in @morrone's jobspec implementation will always have "slot type", which will help reduce the complexity of schedulers.

Good point. For the eventual future where jobspec is emitted by some kind of front end command, this pushes the creation of the "implicit slot" onto the command, unless a user explicitly defines slots through whatever mechanism the command(s) allow. I can't think of any problem with that, and in fact it seems like a good design choice to me.

**Comment by dongahn** [comment]
Ok thanks. I will assume the existence of an explicit slot type in the resources section if a slot is ever specified,
then (unless @morrone disagees).

**Comment by morrone** [comment]
I am on board with removing the implicit slots.  That does also eliminate my concern about needing multiple passes to parse the jobspec.

There is still some question in my mind about whether the jobspec library should assume a default "exclusive" value for users or whether is should just report UNSPECIFIED when it is left unspecified.  Without having three states for "exclusive", there is no way to know if the user actually asked for the state, or whether it was just the default.  And perhaps that might matter.

For instance, if I put "slot" under "node", "node" will be exclusive=false.  In our common production situation, the scheduler will need to ignore that and instead exclusively allocate the nodes (because on most machines that is just our policy).  And in most cases, that will probably be fine because the user didn't really expect to share the node; exclusive=false was just the default value.

On the other hand, what if the user actually _did_ write exclusive=false in their jobspec?  This would seem to say that they explicitly want sharing (perhaps because they want to co-locate some other job on the same node).  Perhaps since this is not allowed by policy on LLNL machines, that should generate an error rather than being overridden by the scheduler.

Maybe that is too obscure to care about for now. :)

**Comment by dongahn** [comment]
Yeah, I think we will eventually need more than a boolean. It seems we will probably need to determine a default action for each cell of a table like below. On each resource type (e.g., core, memory, slot, ...), both sched and jobspec will have an exclusivity property:

|  jobspec(below)/sched-policy(right) | exclusive=yes  | exclusive=false   | go w/ jobspec   |
|---|---|---|---|
|  exclusive=true |  (A) ok |  (B) reject | (C) ok |
|  exclusive=false | (D) reject | (E) ok | (F) ok  |
|  go w/ scheduler policy |  (G) ok | (H) ok   | (I) ?  |

For my current testing, I'm using B and E of the last column.

**Comment by lipari** [comment]
We do allow the user to request exclusive access when the policy is shared (exclusive=false).  So (B) above can be "ok".

**Comment by dongahn** [comment]
@lipari: that would work for node. But the default action may be different if the resource type is higher like rack?

**Comment by dongahn** [comment]
Almost sounds as if we would need more than 3 values then ...

**Comment by dongahn** [comment]
Just to make it clear, the above comment is for the schedulers. The jobspec should be fine with 3 values.

**Comment by morrone** [comment]
Hmm, you mentioned earlier that the implicit slots didn't apply to the named resource, only those below.  But there are examples in there that don't match that.  For instance:

```
version: 1
resources:
  - type: node
    count: 4
tasks:
  - command: hostname
    slot:
      type: node
    count:
      per_slot: 1
attributes:
  system:
    duration: 1 hour
```

Maybe that is just another reason why it is good for us to eliminate implicit slots. :)

Note that the above example will now have to become:

```
version: 1
resources:
  - type: slot
    count: 4
    label: default
    with:
      - type: node
        count: 1
tasks:
  - command: hostname
    slot: default
    count:
      per_slot: 1
attributes:
  system:
    duration: 1 hour
```

The following is not currently valid:

```
version: 1
resources:
  - type: node
    count: 4
    with:
      - type: slot
        count: 1
        label: default
tasks:
  - command: hostname
    slot: default
    count:
      per_slot: 1
attributes:
  system:
    duration: 1 hour
```

**Comment by dongahn** [comment]
> Maybe that is just another reason why it is good for us to eliminate implicit slots. :)

Agreed. 

Sounds great across the board @morrone!

**Comment by grondo** [comment]
If slots become explicit I wonder if we should reevaluate and see if there is a more efficient method to define them in the jobspec... Just an idea, sorry if it is a bit off topic.

**Comment by morrone** [comment]
I could use some help with "Use Case 1.5:: Complex example from OAR".  How were the implicit slots intended to work here?  "type: node" appears twice, so are the slots a concatenation of the nodes?

This would seem to imply that I need slots with the same label, but (perhaps naively) in an early commit I made it explicit that labels must be unique.

Should I change it so that labels need not be unique on slots?  And then describe that slots with the same label are concatenated for purposes of task layout?

This particular example seems like it was always slight off too...where is the "cpu" in the jobspec?

**Comment by grondo** [comment]
> How were the implicit slots intended to work here? "type: node" appears twice, so are the slots a concatenation of the nodes?

This was a tricky one to define, and it would not surprise me if something was a little bit off.

I might not understand, but the "slot" from a task layout point of view is supposed to be created such that we run one copy of flux-broker on each node (perhaps that is obvious). In this case the slots are not all the same shape because the allocation is not homogeneous (That's why `type: node` appears twice I think)

If each non-uniform resource request will require a different slot with a different label, this could make one of common cases of running one thing per node very cumbersome. (Unless a slot of `type: node` could be used to override in the tasks section)

**Comment by grondo** [comment]
> This particular example seems like it was always slight off too...where is the "cpu" in the jobspec?

Yeah, it does seem like the "1 core" part of "1 core on 2 nodes" is missing.. sorry about that

**Comment by grondo** [comment]
> Should I change it so that labels need not be unique on slots? And then describe that slots with the same label are concatenated for purposes of task layout?

Actually, I think I understand what you're saying now and I agree -- we'll need to be able to concatenate slots together and be able to refer to them from a common label for task layout purposes. It would be nice if we could just use a resource type instead of an implicit slot for efficiency, but it does seem like you've found some serious drawbacks with that approach.

Thanks for the detailed attention on this.

**Comment by morrone** [comment]
Part of my confusion was just the wording or the problem in that example, but I see that it is an exact copy of OAR's wording, not something that we came up with.  I don't like their wording. :)  But I think I more-or-less understand what their specification means now.

So here's a question that I can't quite answer from RFC14.  Take this example:

```
version: 1
resources:
  - type: slot
    count: 4
    label: foo
    with:
    - type: node
      count: 1
tasks:
  - command: [ "echo", "$FLUX_TASK_RANK" ]
    slot: foo
    count:
      per_slot: 1
      total: 2
  - command: [ "echo", "$FLUX_TASK_RANK" ]
    slot: foo
    count:
      per_slot: 1
      total: 2
attributes:
  system:
    duration: 1 hour
```

Do we consider this valid?  And what do I see printed out?  Is it "0, 1, 2, 3" or "0, 1, 0, 1" (in random order and on different lines, of course).

If all of the tasks are given a unique task RANK, then one way to avoid needing non-unique labeled slots might be to do the OAR example like so:

```
version: 1
resources:
  - type: cluster
    count: 1
    with:
      - type: slot
        count: 2
        label: foo
        with:
          - type: node
            count: 1
            with:
              - type: memory
                count: 4
                unit: GB
              - type: ib10g
                count: 4
  - type: switch
    count: 1
    with:
      - type: slot
        count: 2
        label: bar
        with:
          - type: node
            count: 1
            with:
              - type: core
                count: 2
tasks:
  - command: [ "flux", "start" ]
    slot: foo
    count:
      per_slot: 1
  - command: [ "flux", "start" ]
    slot: bar
    count:
      per_slot: 1
attributes:
  system:
    duration: 4 hours
```

But that depends on a particular set of semantics around commands in the tasks section, and my guess at those semantics might very well be wrong.

**Comment by dongahn** [comment]
I don't recall we talked much about the task distribution and task grouping semantics of the tasks spawned across different command directives. @grondo may want to weigh in.

**Comment by lipari** [comment]
After reading through the discussion, it is looking more and more like we are trying to cram three independent concepts (resources, slots, and tasks) into two frameworks (resources and tasks).  And while we may be able to successfully do so most of the time, we keep finding cases that are ambiguous or impossible.

I'd like to propose we think of separating this into three groupings:
* The `resource` section would remain the union of all the resources the job requires plus the system duration.  This might ideally be all the scheduler looks for (not sure about this).
* The `slot` section would stipulate the carving out of above resources into subsets for task launching.  A slot could not span multiple nodes.
* The `task` section would stipulate the command(s?) to launch, the mapping to slots, and the number of tasks per slot.

While we may succeed in finding a means to express all of the above problematic examples using just the `resource` and `task` grouping, I suspect it would be very nuanced and difficult for a user to understand and write job specs clearly for.  Creating a dedicated `slot` section would help a user understand the conceptual differences.

This would also help us support MPMD where multiple commands are launched across varying slots under one MPI comm world.

**Comment by lipari** [comment]
... and if the `slot` section were omitted, it would default to 1 core.

**Comment by grondo** [comment]
> Do we consider this valid? And what do I see printed out? Is it "0, 1, 2, 3" or "0, 1, 0, 1" (in random order and on different lines, of course).

a jobspec defines a "program" and each task in a program has a unique rank, and yes your example looked valid to me (but at this point I'd want one more opinion on that :smile:)

I'm still digesting the second part of your post.

**Comment by grondo** [comment]
@lipari, I don't necessarily disagree with your observation, however I don't think I actually see how you could define arbitrary slots in their own section. Really they are just a grouping of resources (and they started out that way IIRC) I could be missing what you're getting at though.

**Comment by dongahn** [comment]
@lipari: Thank you for sharing your observation. 

After playing with jobspec in resource matching logic, here is my take on it. Please consider it as lessons learned from a consumer of jobspec. 

> - The resource section would remain the union of all the resources the job requires 

Agreed.

> plus the system duration. This might ideally be all the scheduler looks for (not sure about this).

Disagreed. 

Time (e.g., duration, start_time, begin_after...) and space should cleanly be separated. I like the way the jobspec currently makes this separation. 

In fact, I plan to make this separation for my match callback interfaces as well. 

I really don't want resource selection match policy plugin writers to worry about time. They should only concern wether a evaluating resource vertex has all of the required properties and score it. The rest of considerations including time should be taken care by the scheduling infrastructure. I am pretty confident I will be able to do this at this point.

> - The slot section would stipulate the carving out of above resources into subsets for task launching. A slot could not span multiple nodes.

I am not sure if I understand your proposal, but this may be similar to implicit slotting that appears in the tasks section? As @morrone, @grondo agreed, this may belong to a different interface than "canonical" interface? From the scheduler's point of view, having a guarantee on the existence of `slot` significantly simplifies things

- The task section would stipulate the command(s?) to launch, the mapping to slots, and the number of tasks per slot.

Agreed. I believe primitives are there. It is just that we need to write down various semantics somewhere?

**Comment by dongahn** [comment]
> ... and if the slot section were omitted, it would default to 1 core.

Again, this is a "canonical" spec. I would say `slot` should not be omitted if ever need to be specified.

**Comment by lipari** [comment]
@grondo and @dongahn, here is the general idea for specifying Case 1.5.  Note that all the resources are defined in the `resources` section, but then these are carved down to labeled slots in the `slots` section.  The `tasks` section becomes a simple mapping of the command to the labeled slots.

I'm not sure I got it exactly right, but the general idea is to independently define slots in their own section instead of weaving slot definitions into the `resources` or `tasks` sections.

```
version: 1
resources:
  - type: cluster
    count: 1
    with:
    - type: node
      count: 2
       with:
        - type: memory
          count: 4
          unit: GB
          - type: ib10g
            count: 1
    - type: switch
      count: 1
      with:
      - type: node
        count: 2
        with:
          - type: core
            count: 1

slots:
    - label: foo
      count: 2
      with:
          - type: node
            count: 1
            with:
              - type: memory
                count: 2
                unit: GB
              - type: ib10g
                count: 1
 
    - label: bar
      count: 2
        with:
        - type: switch
          count: 1
          with:
          - type: node
            count: 1
            with:
              - type: core
                count: 1

tasks:
 - command: [ "flux", "start" ]
    count: 2
      slot: foo
 - command: [ "flux", "start" ]
    count: 2
      slot: bar

attributes:
  system:
    duration: 4 hours
```

**Comment by dongahn** [comment]
@lipari: From a scheduler developer's point of view, I won't like it unless jobspec implementation will hide this and merge the slot back into the resources section.

**Comment by grondo** [comment]
and something would have to verify that the slots are a subset of the resources section anyway.

Conceptually, slots are just labeled grouping of requested resources, I'm not sure how to efficiently split that from the resource section.

If explicit slots are always *required* (e.g. we do not allow named resource types to implicitly be a slot) then really the `resources` section becomes a "slot definition" section where you define the shape of the slots you need and the number of each slot that you are requesting.

**Comment by lipari** [comment]
@grondo, I agree with your point that creating a `slots` section introduces a need to verify that all of the slots will fit into the resources defined in the `resources` section.  So, it's not the perfect solution.

What it does do however, IMO, is create a simpler model for the user to understand and follow.  Seeing labeled slots defined on their own, allows both a simplification of the `resources` and `tasks` section and eliminates the confusion of implicit vs explicit definitions.

The default definition of a slot is one core (and maybe its portion of the NUMA's memory).  So the notion of requiring a `tasks` section is optional.

Finally, I still see this as making the scheduler's work easier.  As I understand what the scheduler does based on the prototype I created, it is only looking for a complete set of resources presented in the `resources` section.  The task mapping to slots is a spec that is passed on the execution engine.

**Comment by grondo** [comment]
> What it does do however, IMO, is create a simpler model for the user to understand and follow. 

Keep in mind that this is just the canonical jobspec, and it is unlikely in the long term that users would ever write directly to this spec. On the other side, I think that *clients* of the jobspec API (like sched and exec subsystems) likely won't be looking at the raw yaml representation, but would have some "nicer" API (e.g. I would assume that sched could look at resources section with slot information masked out and get what you want)

That being said, I think I defer further feedback to @morrone and @dongahn who are actually working on the problem.

**Comment by morrone** [comment]
> If explicit slots are always required (e.g. we do not allow named resource types to implicitly be a slot) then really the resources section becomes a "slot definition" section where you define the shape of the slots you need and the number of each slot that you are requesting.

Yes, that is exactly right.  That is what I was feeling out in the second and third examples of [this comment](https://github.com/flux-framework/rfc/issues/101#issuecomment-329029699).

In rewriting the examples to match the explicit-slot-only requirement, I kept changing the major count from something like a node to a slot.  I wasn't sure how I felt about that at first, but I'm starting to come around to it being positive.

My main question at the moment is what to do about this rule:

> and each label in the current jobspec must be unique. 

Should I retract that?  Should we allow multiple slots to have the same name label?

**Comment by dongahn** [comment]
> (e.g. I would assume that sched could look at resources section with slot information masked out and get what you want)

From talking to @morrone, sched does need to look at `slot` info for valid grouping. Plus, having `slot` appears before its resource subtree in DF traversal would make it easy to propagate exclusive bit down to the subtree resource selection (currently the entire subtree being exclusive)

> That being said, I think I defer further feedback to @morrone and @dongahn who are actually working on the problem.

I think I understand where @lipari is coming from -- human-user-facing point of view. But my general position is to keep the canonical jobspec to be rather machine-friendly for a lack of better words. 

This concept is similar to programming languages and compilers. While a goal of a higher-level language should be user-friendliness and user productivity, a source code written in that language will be ultimately all translated down to some canonical forms (called "lowering") to make it easy for the next layers to work w/ -- e.g., intermediate representation (IR) or even further lower level machine code representation. 

A common, ease-to-work-with, explicit representation would make other layers be layered much better and perform, while leaving user friendliness to some other interfaces.

**Comment by grondo** [comment]
> Should I retract that? Should we allow multiple slots to have the same name label?

The answer isn't clear to me. It seems really useful for the case of starting an instance where you want to launch a copy of `flux-broker` on each node, even if each node had a different set of resources under it and thus had a different shape slot. 

However, it does seem to allow more cases of ambiguity. For example if you had 2 differently shaped slots with the same label, then ran two different commands on each results would be unspecified? Maybe that is ok. It would be the same as in the implicit slot case if you ran two different commands with `slot: type: node`

**Comment by dongahn** [comment]
If there is a common use case like this, we should allow it and document this undefined behavior  in this RFC and/or task execution RFC?

**Comment by morrone** [comment]
Going back to grondo's comment earlier:

> a jobspec defines a "program" and each task in a program has a unique rank

That is good.  That is how I've been thinking about it, and that is how I hoped it would work.  But there is a bit of ambiguity in my mind surround in the semantics in the "tasks" section.

The tasks sections allows unlimited different programs to be listed.  This seems to largly be aimed at supporting MPMD.  However if all of the tasks are then given unique sequential IDs and are treated as SPSD/SPMD, it is not clear to me that the semantics there will work out sanely.  If we want to treat it as a single program, then maybe we should not permit many different commands to be started.  Maybe there should be a single command in the tasks section, and optionally allow different arguments to support SPMD.  Maybe MPMD requires starting a new flux instance, and submitting separate jobspecs to the new flux instance for each of the commands in MPMD.

**Comment by dongahn** [comment]
> Maybe MPMD requires starting a new flux instance, and submitting separate jobspecs to the new flux instance for each of the commands in MPMD.

Just a quick comment:

Talking to users, the diversity of "MPMD" is getting increasingly larger. 

Traditionally, MPMD has often referred to tightly coupled applications like CCSM (5 different climate executables coupled through MPI). But I see increasing needs for more loosely coupled applications (simulation application + analysis engine; simulation application + in-situ visualizer; simulation + IO accelerators...). 

Also their co-location and communication requirements are also getting diverse. In some mode, the distinct job elements need to be completely co-located; in others, they simply partition allocated resources; or somewhere in-between. 

Depending on the MPMD communication models and mechanisms (e.g., MPI, shared file systems, ...) among these distinct job elements, it can make a separate-flux-instance approach significantly difficult to work with. (Although they would be equally many cases that this approach would work fine). 

So, I would be cautious about getting rid of MPMD support within a single jobspec.

**Comment by grondo** [comment]
@dongahn, you may have a new use case we weren't considering before here. I don't have a preconceived conclusion here, but maybe there is something higher level than a jobspec that could handle these very loosely coupled cases (we might take care to not take too many features into the jobspec), as @morrone alluded to above?

@morrone, for simpler MPMD that are part of a single COMM_WORLD or equivalent concept in another paradigm, is there a requirement that task ranks can't be unique among all spawned tasks?
I might not have the full picture on how MPMD works, but I was treating as 1 program with various executables.

**Comment by grondo** [comment]
In terms of Slurm, the current `tasks` section allows the jobspec to express things that users might use `srun --multi-prog` for now.  (with many improvements however)

What @dongahn mentioned sounds something like ["job packs"](https://slurm.schedmd.com/SLUG16/Job_Packs_SUG_2016.pdf) which (for some use cases at least) seem like a good candidate for a higher level approach mentioned by @morrone.

**Comment by dongahn** [comment]
> @dongahn, you may have a new use case we weren't considering before here. I don't have a preconceived conclusion here, but maybe there is something higher level than a jobspec that could handle these very loosely coupled cases (we might take care to not take too many features into the jobspec), as @morrone alluded to above?

Maybe. In general, I think it will be beneficial to go through some key examples.

When I surveyed users for ECP proposal write-up a year ago, two things were identified as important: 1) UQ & V&V type load and 2) better support for distinct job elements. 

From talking to them, though, if 1 program with various executables is supported in jobspec, I thought we are close to a reasonable support level. 

I think this is one of the co-design activities (e.g., next-gen. IO team and workflow team and whoever has traditional MPMD).

**Comment by grondo** [comment]
Just reading through the job packs presentation, it seems like there may be use cases for running MPMD as a single MPI_COMM_WORLD, or optionally with each "step" in a separate COMM_WORLD. 
I'll leave this up to @morrone, but maybe questions about the MPMD semantics of the tasks section should be moved to another issue?

**Comment by morrone** [comment]
>  for simpler MPMD that are part of a single COMM_WORLD or equivalent concept in another paradigm, is there a requirement that task ranks can't be unique among all spawned tasks?

I believe that it is technically possible to have different executables wind up in the same program, but in practice I don't think it is something that is done.  The problem is that all of the executables would have to follow very tightly coordinated and agreed upon behaviours at start up in order to handle configuration options, generate other comm sets, and begin working in a less globally synchronous manner.

Technically speaking you _could_ copy that same exact startup code into multiple executables, but  I believe that more generally people tend to have one coordinating executable that they launch.  That one executable then loads different modules and coordinates getting their comm requirements in place and such.

But I don't claim full knowledge of the space.

It would be nice to know how people use --multi-prog in slurm.  My guess is that when the programs are different, they do not join into a shared MPI comm space.

My biggest concern is jut understanding how semantics if we keep multiple command support in the spec.  What is are the semantics of multiple commands sharing a slot?  Which component(s) is(are) responsible for making sense of that?

**Comment by grondo** [comment]
> My biggestt concern is jut understanding how semantics if we keep multiple command support in the spec. What is are the semantics of multiple commands sharing a slot? Which component(s) is(are) responsible for making sense of that?

That seems like a valid concern. Should we move this to another issue or are you ok discussing it here?

**Comment by dongahn** [comment]
> I believe that it is technically possible to have different executables wind up in the same program, but in practice I don't think it is something that is done. 

FWIW, I do recall many years back (supporting user debugging), CCSM users were doing exactly this, and ANL also required this capability as part of a build SOW as well. -- This is >10 years ago. 

In any case, as far as we have reasonable baseline MPMD support with clear semantics at the jobspec level, which then is used as our building to support higher level workflow, I think this should be fine. 

Once we have reasonable draft on this subject, I will be taking on an action to talk to a few users + IO + WF folks about it. Since I suggested "co-desing" as part of Schulz's L2 anyway.

**Comment by morrone** [comment]
I'm OK with discussing it here for now.  The results of that discussion will help make a decision about whether or not slot labels should be unique, which helps me write the spec update that deals with removing the implicit slots.

**Comment by grondo** [comment]
> What is are the semantics of multiple commands sharing a slot? Which component(s) is(are) responsible for making sense of that?

I'm not sure I follow what you mean by "multiple commands sharing slot"? Is the case you are worried about where a user wants to specify one command that runs in the same slot as some other command? There might need to be an extra keyword for co-scheduling tasks like that I guess.

**Comment by grondo** [comment]
Here's another parting thought for the day.  For a simple `tasks` section like the following:

```yaml
tasks:
 - command: [ "flux", "start" ]
    count: 4
    slot: default
```

This could be expanded equivalently as

```yaml
tasks:
 - command: [ "flux", "start" ]
    count: 1
    slot: default
 - command: [ "flux", "start" ]
    count: 1
    slot: default
 - command: [ "flux", "start" ]
    count: 1
    slot: default
 - command: [ "flux", "start" ]
    count: 1
    slot: default
```

For the multi-command case, some of the `command` arrays just happen to be different. So this case, although simple, seems very straightforward.

**Comment by morrone** [comment]
Yes, I suppose that is just another bit of validation that should happen.  And maybe we can figure out most of hte rules later.  For instance, is it an error to ask for more slots than are used by the tasks section?  Is it an error to over use the slots?  For instance, what if there was a fifth command in the previous comment's example?  Probably not...but working out the semantics for where things actually get run will need to be decided.

But that can probably wait for another ticket.

**Comment by grondo** [comment]
> For instance, is it an error to ask for more slots than are used by the tasks section? Is it an error to over use the slots?

Yeah, these are good questions. IMO, It would be great if the first was flagged by a validator/lint tool as a warning (though I'm not sure why it wouldn't be allowed) and the second should be an error during validation unless some kind of overcommit flag was enabled?

It would be cool if we made an issue or wiki page with a table or list of these cases.
